### PR TITLE
Twitter Oauth Login

### LINKED ISSUE #5363: Implement additional signups
Please implement additional signups as follows for speakers and attendees:

* [ ] If you would like your information to be shared with potential recruiters, please check this box. By checking, you allow the Linux Foundation to share your information with potential recruiters. 

* [ ] By checking here, you opt-in to being listed in the event's online attendee directory. Some of your registration data will be made available to other event attendees in the directory (name, title, and company name only). 

* [ ] By checking here, you opt-in to receiving communications from the event's sponsors regarding their own news and messages (sent by The Linux Foundation; your contact information will not be shared).

* [ ] By checking here, you opt-in to receiving occasional communications from The Linux Foundation relating to our other open source projects, events, and services like training and certifications. If so, you can unsubscribe from email communications at any time using the email's Unsubscribe link. Please note that The Linux Foundation may contact you regarding events and other services you actually register for, even if you do not check this box or if you use a marketing email's Unsubscribe link.

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #5363 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:

- Added Twitter Oauth Endpoint 
- Added functionality to save in database
- Added new fields for twitter login hash and twitter id

### CODE REVIEW

**Comment by prateekj117** [comment]
@mrsaicharan1 There is already an issue regarding this. #5175 and I was working on it. Maybe you can take up the twitter OAuth as you have already started working on it, and I can work on the instagram OAuth.

**Comment by mrsaicharan1** [comment]
@prateekj117 I think that would be a good idea! You can go head with the Instagram Oauth.

**Comment by mrsaicharan1** [comment]
I've implemented the Twitter OAuth using Flask-Dance, can I get a review on this just to see if everything's ok ?

**Comment by iamareebjamal** [comment]
Fix hound issues

**Comment by mrsaicharan1** [comment]
@iamareebjamal The hound tests are passing. I guess there's less coverage for the code.

**Comment by iamareebjamal** [comment]
OK, what happens in the case I have not registered the application ID and secret in the app

**Comment by mrsaicharan1** [comment]
@iamareebjamal I've checked for the absence of credentials so that it returns a response of `keys not found`

**Comment by iamareebjamal** [comment]
Fix the build

**Comment by mrsaicharan1** [comment]
@iamareebjamal Fixed. Please review.

**Comment by iamareebjamal** [comment]
Remove empty migration and use Flask Dance

**Comment by mrsaicharan1** [comment]
@iamareebjamal Implemented Twitter Oauth using Flask-Dance itself. Which empty migration file? Are you talking about the merged migration file `migrations/versions/96860fd619d7_.py`. Removing that would occur in an error *Multiple migrations heads*. What would be the best way to do this?

File: app/api/auth.py [hunk 1]
```diff
@@ -138,6 +140,24 @@ def login_user(provider, auth_code):
...
+        twitter_blueprint = make_twitter_blueprint(api_key=payload['client_id'], api_secret=payload['client_secret'])
+        app.register_blueprint(twitter_blueprint, url_prefix='/twitter_login')
+        if not twitter.authorized:
+            return redirect(url_for('twitter.login'))
+        account_info = twitter.get('account/settings.json')
+        if account_info.ok:
+            account_info_json = account_info.json
+        user = User()
+        user.first_name = account_info_json['name']
+        user.twitter_id = account_info_json['screen_name']
```
Comment (uds5501) [inline app/api/auth.py]: @mrsaicharan1 you are using a `db.BigInteger` column to store a string(`screen_name`). Is this valid?

File: app/api/auth.py [hunk 2]
```diff
@@ -138,6 +140,24 @@ def login_user(provider, auth_code):
...
+        user.twitter_login_hash = random.getrandbits(128)
+        user.password = ''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(8))
+        save_to_db(user)
+        return make_response(jsonify(user_id=user.id, twitter_login_hash=user.twitter_login_hashpython3,
```
Comment (uds5501) [inline app/api/auth.py]: @mrsaicharan1 Also, I believe that `user.twitter_login_hashpython3` is a typo, you should fix it

File: migrations/versions/3b80e4ac0ecb_.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '3b80e4ac0ecb'
+down_revision = 'e3caa0f2a16c'
```
Comment (iamareebjamal) [inline migrations/versions/3b80e4ac0ecb_.py]:
```suggestion
down_revision = ('e3caa0f2a16c', '91ee86a38001')
```

Try this and delete the empty migration.

File: app/api/auth.py (hunk 2 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: Thanks for identifying. It's not a typo though, need to add it in the Alchemy model.

File: requirements/common.txt [hunk 4]
```diff
@@ -6,6 +6,7 @@ Flask-Migrate>=2.0.0,<3
 Flask-Login>=0.4,<1
 Flask-Scrypt>=0.1.3.6,<0.2
 Flask-JWT>=0.3.2,<0.4
+Flask-Dance
```
Comment (iamareebjamal) [inline requirements/common.txt]: Pin the version

Comment (mrsaicharan1) [reply requirements/common.txt]: @iamareebjamal Added

File: app/api/auth.py [hunk 5]
```diff
@@ -138,6 +140,24 @@ def login_user(provider, auth_code):
             'client_id': provider_class.get_client_id(),
             'client_secret': provider_class.get_client_secret()
         }
+        if not payload['client_id'] and payload['client_secret']:
```
Comment (iamareebjamal) [inline app/api/auth.py]: Wrong condition

Comment (mrsaicharan1) [reply app/api/auth.py]: Yes, this is wrong. I was just waiting for the #5589 to get merged so that I can import the helper function implemented there.

Comment (mrsaicharan1) [reply app/api/auth.py]: NotImplementedError

File: app/api/auth.py (hunk 2 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: @uds5501 Took care of it. Thanks for spotting it.

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @iamareebjamal There was a small bug related to flask dance setup. Now everything works fine. @shreyanshdwivedi has the front end been set up for Twitter Oauth?

**Comment by uds5501** [comment]
@mrsaicharan1 How are you testing the setup? I could use some guidance regarding this for github's OAuth implementation

**Comment by mrsaicharan1** [comment]
@uds5501 here's the manual testing which I performed.
1. Submit the Oauth tokens and check if they are getting saved onto the database by placing print statement stubs in the code.
2. Further, check if you're authorized or not by redirecting to `github.login` route. This should take you to the authorization permissions page. 
3. After this, you should be redirected to the callback url.
The response looks something like this, for twitter that is.
`{
  "twitter_id": "reddy_reddy1", 
  "twitter_login_hash": 165833506533213567384652371511776360413, 
  "user_id": null
}`

**Comment by uds5501** [comment]
@mrsaicharan1 Alright, thank you!

**Comment by shreyanshdwivedi** [comment]
@mrsaicharan1 waiting for this to get merged. I'll be do it as soon as this gets merged

**Comment by mrsaicharan1** [comment]
@shreyanshdwivedi Having multiple migrations. Fixing them up right now.

**Comment by mrsaicharan1** [comment]
I've deleted the previous migration and set the desired migration accordingly. Still says that there's multiple migration heads. What might be the problem?

**Comment by mrsaicharan1** [comment]
Will be opening a new branch due to a migration fault

**Comment by iamareebjamal** [comment]
PRs should not be closed. Just add a new commit on the same branch

**Pull request closed without merging**